airflow/jobs/scheduler_job_runner.py

```diff
@@ -966,6 +995,157 @@ def _update_dag_run_state_for_paused_dags(self, session: Session = NEW_SESSION)
         except Exception as e:  # should not fail the scheduler
             self.log.exception("Failed to update dag run state for paused dags due to %s", e)
 
+    @provide_session
+    def _end_active_spans(self, session: Session = NEW_SESSION):
+        # No need to do a commit for every update. The annotation will commit all of them once at the end.
+        for key, span in self.active_spans.get_all().items():
+            from airflow.models.taskinstance import TaskInstanceKey
+
+            if isinstance(key, TaskInstanceKey):  # ti span.
+                # Can't compare the key directly because the try_number or the map_index might not be the same.
+                ti: TaskInstance = session.scalars(
+                    select(TaskInstance).where(
+                        TaskInstance.dag_id == key.dag_id,
+                        TaskInstance.task_id == key.task_id,
+                        TaskInstance.run_id == key.run_id,
+                    )
+                ).one()
+                if ti.state in State.finished:
+                    self.set_ti_span_attrs(span=span, state=ti.state, ti=ti)
+                    span.end(end_time=datetime_to_nano(ti.end_date))
+                    ti.span_status = SpanStatus.ENDED
+                else:
+                    span.end()
+                    ti.span_status = SpanStatus.NEEDS_CONTINUANCE
+            else:
+                dag_run: DagRun = session.scalars(select(DagRun).where(DagRun.run_id == key)).one()
+                if dag_run.state in State.finished_dr_states:
+                    dagv = session.scalar(select(DagVersion).where(DagVersion.id == dag_run.dag_version_id))
+                    dag_run.set_dagrun_span_attrs(span=span, dagv=dagv)
+
+                    span.end(end_time=datetime_to_nano(dag_run.end_date))
+                    dag_run.span_status = SpanStatus.ENDED
+                else:
+                    span.end()
+                    dag_run.span_status = SpanStatus.NEEDS_CONTINUANCE
+                    initial_dag_run_context = Trace.extract(dag_run.context_carrier)
+                    with Trace.start_child_span(
+                        span_name="current_scheduler_exited", parent_context=initial_dag_run_context
+                    ) as s:
+                        s.set_attribute("trace_status", "needs continuance")
+
+        self.active_spans.clear()
+        self.active_spans.clear()
```

Comment:
Is this meant to duplicate or is that a typo?

Reply:
This must have been put there accidentally after one of the many merges with main. I cleaned it up.